pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/BasePartitionUpsertMetadataManager.java

```diff
@@ -535,6 +537,13 @@ public GenericRow updateRecord(GenericRow record, RecordInfo recordInfo) {
 
   protected abstract GenericRow doUpdateRecord(GenericRow record, RecordInfo recordInfo);
 
+  @Override
+  public boolean shouldDropRecord(RecordInfo recordInfo) {
+    return doShouldDropRecord(recordInfo);
```

Comment:
doShouldDropRecord() seems not necessary, as shouldDropRecord() does nothing but call this method.

how about to move the changes from L307-L326 in `ConcurrentMapPartitionUpsertMetadataManager` to this shouldDropRecord() method and add a `handleOutOfOrderEvent` method in base class to let subclass to override this part only.

Reply:
> how about to move the changes from L307-L326 in ConcurrentMapPartitionUpsertMetadataManager to this shouldDropRecord() method

@klsince `ConcurrentMapPartitionUpsertMetadataManager` tries to access the ConcurrentHashMap for previous record for comparing comparison values. We will not have access to that HashMap here. 

> add a handleOutOfOrderEvent method in base class

We already have a `handleOutOfOrderEvent` method in base class just below this method which updates metrics.

Let me know if I am missing anything.